watch/watch_test.go

```diff
@@ -0,0 +1,106 @@
+package watch
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+	"time"
+
+	devtest "github.com/puma/puma-dev/dev/devtest"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	tmpDir        = filepath.Join(devtest.ProjectRoot, "tmp")
+	tmpFilename   = "restart.txt"
+	tmpRestartTxt = filepath.Join(tmpDir, tmpFilename)
+)
+
+type Notice struct{}
+
+func TestMain(m *testing.M) {
+	if err := fileCreateAll(tmpRestartTxt); err != nil {
+		panic(err)
+	}
+
+	result := m.Run()
+
+	os.RemoveAll(tmpDir)
+
+	os.Exit(result)
+}
+
+func TestWatch_ExpectTimeout(t *testing.T) {
+	watchTriggered := watchTmpFileWithTimeout(t, func() {})
+
+	assert.False(t, watchTriggered)
+}
+
+func TestWatch_ExpectTouchSignalAfterModify(t *testing.T) {
+	watchTriggered := watchTmpFileWithTimeout(t, func() {
+		touchFile(t, tmpRestartTxt)
+	})
+
+	assert.True(t, watchTriggered)
+}
+
+func fileCreateAll(fullPath string) error {
+	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
+		return err
+	}
+
+	f, err := os.Create(fullPath)
+
+	if err != nil {
+		return err
+	}
+
+	f.Close()
+
+	return nil
+}
+
+func touchFile(t *testing.T, fullPath string) {
+	if err := exec.Command("sh", "-c", fmt.Sprintf("touch %s", fullPath)).Run(); err != nil {
+		t.FailNow()
+	}
+}
+
+func watchTmpFileWithTimeout(t *testing.T, f func()) bool {
+	watchDone := make(chan struct{})
+	watchTriggered := false
+
+	if !devtest.FileExists(tmpRestartTxt) {
+		assert.FailNow(t, fmt.Sprintf("%s does not exist", tmpRestartTxt))
+	}
+
+	go func() {
+		err := Watch(tmpRestartTxt, watchDone, func() {
+			watchTriggered = true
+			watchDone <- Notice{}
+		})
+
+		if err != nil {
+			assert.FailNow(t, err.Error())
+		}
+	}()
+
+	timeoutDone := make(chan struct{})
+	go func() {
+		time.Sleep(2 * time.Second)
+		timeoutDone <- Notice{}
+	}()
```

Comment:
we use this separate channel and go routine to wait for 2 seconds (latency on fsevents is set to .5s). if the touch was going to trigger the `Watch`, it would happen before 2s. 

so, we can ensure `touch` triggers and no touch doesn't trigger (i.e. times out)